src/mecca/reactor/platform/kqueue.d

```diff
@@ -0,0 +1,398 @@
+                        continue;
+                    break;
+                case Write:
+                    if (event.filter != EVFILT_WRITE)
+                        continue;
+                    break;
+                case Both:
+                    assert(false);
+                }
+
+                auto state = &ctx.states[dir];
+                with(FdContext.Type) final switch(state.type)
+                {
+                case None:
+                    if( cast(Direction)dir==Read || event.filter != EVFILT_READ ) {
+                        // Since most FDs are available for write most of the time, almost any wakeup would trigger
+                        // this warning. As such, we log only if one of two conditions are met:
+                        // Either we got this condition on a read, or we got this condition on a write, but the FD is
+                        // not read ready.
+                        WARN!"kqueue64s for returned fd %s events %s which is not listening for %s"(
+                                ctx.fdNum, event.filter, dir);
+                    }
+                    break;
+                case FiberHandle:
+                    theReactor.resumeFiber(state.fibHandle);
+                    break;
+                case Callback:
+                    state.callback(state.opaq);
+                    break;
+                case CallbackOneShot:
+                    state.type = None;
+                    state.callback(state.opaq);
+                    break;
+                case SignalHandler:
+                    ASSERT!"Event signal %s was not the same as the registered signal %s"(event.ident == ctx.fdNum, event.ident, ctx.fdNum);
+                    state.signalHandler(toOSSignal(event.ident));
+                    break;
+                }
+            }
+        }
+
+        return true;
+    }
+
+private:
+
+    void internalRegisterFD(int fd, FdContext* ctx, Direction dir) @trusted @nogc
+    {
+        import mecca.reactor.platform : EV_ADD, EV_CLEAR, EV_ENABLE;
+
+        ASSERT!"ctx is null"(ctx !is null);
+        static immutable short[2] filters = [EVFILT_READ, EVFILT_WRITE];
+
+        foreach (filter ; filters)
+        {
+            const kevent64_s event = {
+                ident: fd,
+                filter: filter,
+                flags: EV_ADD | EV_ENABLE | EV_CLEAR,
+                udata: cast(ulong) ctx
+            };
+
+            const result = queueEvent(event);
+            errnoEnforceNGC(result >= 0, "Adding fd to queue failed");
+        }
+    }
+
+    void internalDeregisterFD(int fd, FdContext* ctx, bool fdIsClosing) nothrow @trusted @nogc
+    {
+        import core.stdc.errno : errno;
+        import mecca.reactor.platform : EV_DELETE;
+
+        // Events are automatically removed when a file descriptor is closed.
+        // This will save us one system call.
+        if (fdIsClosing)
```

Comment:
Will this not be a problem if a new fd is created with the same number immediately after?

Reply:
Immediately after what?

Reply:
e.g:
1. deregisterFd closing=true
2. close fd
3. Open new file/pipe, get same fd that was just closed, before epoll/kqueue gets to run
4. Next poller call will incorrectly still be registered, but on the new fd?
```

Reply:
I don't know how the internals work but I would assume the operating system places an event to the queue to remove that file descriptor from the queue. In that case it will be removed the next time kqueue is run. The man page says:

```
Calling close() on a file descriptor will remove any kevents that reference the descriptor.
```

Reply:
Sounds like it is fine then.